### PR TITLE
Try Windows again

### PR DESCRIPTION
**Pull request by ararslan**
Modeled after Simon Byrne's PR for libRmath-julia.

### CODE REVIEW

**Comment by ararslan** [comment]
Okay, this gets the build working, but the Julia test that ensures that the file exists says it isn't, so the tests fail.

**Comment by ararslan** [comment]
It makes libfftwf3.dll but not libfftw... so single precision but no double. The makefile must be wrong.

File: test.jl [hunk 1]
```diff
@@ -0,0 +1,23 @@
+using Base.Test
+
+for (root, dirs, files) in walkdir(ENV["APPVEYOR_BUILD_FOLDER"])
```
Comment (ararslan) [inline test.jl]: haha this ends up listing all of the FFTW source files too

**Comment by ararslan** [comment]
Adding `--enable-double` didn't seem to do it either...

**Comment by ararslan** [comment]
This is only producing the single precision binaries. Am I doing something obviously wrong here, @tkelman? The Makefile here seems to mostly match fftw.mk from Base as of the v0.6.0 tag, unless I'm missing something.

**Comment by ararslan** [comment]
Haaaaaaaaaaaah, `--enable-single` produces _only_ single precision binaries. So I guess I have to run the single and double builds separately since they require different configuration flags.

**Comment by ararslan** [comment]
Linux failure is because it can't find the Julia that gets installed, Windows failure is because it's producing .dll.a instead of .dll. What in the world is .dll.a??

**Comment by ararslan** [comment]
And on macOS:
```
things work: Error During Test
  Got an exception of type TypeError outside of a @test
  TypeError: in anonymous, in ccall: first argument not a pointer or valid constant expression, expected Ptr, got Tuple{Symbol,String}
  Stacktrace:
   [1] macro expansion at /Users/travis/build/ararslan/fftw-builder/test.jl:27 [inlined]
   [2] macro expansion at ./test.jl:943 [inlined]
   [3] anonymous at ./<missing>:?
   [4] include_relative(::Module, ::String) at ./loading.jl:464
   [5] include(::Module, ::String) at ./sysimg.jl:14
   [6] process_options(::Base.JLOptions) at ./client.jl:315
   [7] _start() at ./client.jl:383
Test Summary: | Error  Total
things work   |     1      1
ERROR: LoadError: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
while loading /Users/travis/build/ararslan/fftw-builder/test.jl, in expression starting on line 26
```
Maybe something is wrong with the compiled dylib? The code that's failing was copied directly from `__init__()` in FFTW.jl.

**Comment by ararslan** [comment]
Great, the new state of things:
* 32-bit Linux:
  - Can't find the Julia binary
* 64-bit Linux, Windows, macOS
  - Something's wrong with `ccall`

Edit: I'm just bad at `ccall`. Needed some `const`s in there.

**Comment by ararslan** [comment]
IT WOOOOOOOOOORRRRRKKKSSSSSSSS!!! Now just need to figure out the 32-bit Linux Julia installation thing.

File: Makefile [hunk 2]
```diff
@@ -3,7 +3,24 @@ SRCDIR := $(BUILDDIR)/src
...
+THREADS := _threads
+ifeq ($(shell uname),Darwin)
+SHLIB_EXT := dylib
+else
+SHLIB_EXT := so
+endif
+endif
+
+CYGWIN_single := cygfftw3f-3.dll
+CYGWIN_double := cygfftw3-3.dll
```
Comment (tkelman) [inline Makefile]: should not be using this, wrong compiler. missing --host when you configured

**Comment by ararslan** [comment]
```
checking host system type... Invalid configuration `i686-w64-mingw32-gcc': machine `i686-w64-mingw32' not recognized
configure: error: /bin/sh ./config.sub i686-w64-mingw32-gcc failed
make: *** [Makefile:72: /cygdrive/c/projects/fftw-builder/build/lib/libfftw3f-3.dll] Error 1
make: Leaving directory '/cygdrive/c/projects/fftw-builder'
```
Worked passing `CC` directly though. Any idea why?

File: Makefile [hunk 3]
```diff
@@ -41,6 +39,7 @@ FFTW_CONFIG += --enable-sse2 --enable-fma
 endif
 
 ifeq ($(OS),Windows_NT)
+CONFIG += --host=$(ARCH)-w64-mingw32-gcc
```
Comment (tkelman) [inline Makefile]: no -gcc in this, it's a toolchain prefix, not a compiler name

Comment (ararslan) [reply Makefile]: Thanks, changed. Any idea about the 32-bit Linux "can't find Julia" error?

Comment (tkelman) [reply Makefile]: what's the exact error? probably need to install the 32 bit libc

Comment (ararslan) [reply Makefile]:
```
/home/travis/.travis/job_stages: line 57: /home/travis/julia/bin/julia: No such file or directory

The command "julia -E 'versioninfo()'" failed and exited with 127 during .
```
https://travis-ci.org/ararslan/fftw-builder/jobs/261622600

Comment (tkelman) [reply Makefile]: run ldd on it. that doesn't necessarily mean julia itself isn't found, it may mean one of its dependencies

Comment (ararslan) [reply Makefile]: Fails with "not a dynamic executable"

Comment (ararslan) [reply Makefile]: Works on 64-bit

**Comment by ararslan** [comment]
Aha! The issue with 32-bit Julia is that it was indeed missing some 32-bit system libraries. Installing 32-bit GCC, Make, and zlib seems to have done the trick. Thanks so much, Tony!

**Comment by tkelman** [comment]
shouldn't need make or a toolchain in the test environment - in fact better to test without them since some users won't have them. 32 bit libz and libc should be enough

**Comment by ararslan** [comment]
How do I get 32-bit libc from Apt without getting all of GCC?

**Comment by ararslan** [comment]
32-bit Linux doesn't seem to like the binary built with g++ -m32:
```
=====
Full contents of build/lib:
libfftw3.a
libfftw3f.a
libfftw3f.la
libfftw3f.so
libfftw3f.so.3
libfftw3f.so.3.5.6
libfftw3f_threads.a
libfftw3f_threads.la
libfftw3f_threads.so
libfftw3f_threads.so.3
libfftw3f_threads.so.3.5.6
libfftw3.la
libfftw3.so
libfftw3.so.3
libfftw3.so.3.5.6
libfftw3_threads.a
libfftw3_threads.la
libfftw3_threads.so
libfftw3_threads.so.3
libfftw3_threads.so.3.5.6
pkgconfig
=====
Full contents of build/bin:
fftwf-wisdom
fftw-wisdom
fftw-wisdom-to-conf
=====
Looking for:
/home/travis/build/ararslan/fftw-builder/build/lib/libfftw3_threads.so
/home/travis/build/ararslan/fftw-builder/build/lib/libfftw3f_threads.so
Test Summary: | Pass  Total
things exist  |    2      2
ERROR: LoadError: error compiling anonymous: could not load library "/home/travis/build/ararslan/fftw-builder/build/lib/libfftw3_threads.so"
libfftw3.so.3: cannot open shared object file: No such file or directory
Stacktrace:
 [1] include_relative(::Module, ::String) at ./loading.jl:464
 [2] include(::Module, ::String) at ./sysimg.jl:14
 [3] process_options(::Base.JLOptions) at ./client.jl:315
 [4] _start() at ./client.jl:383
while loading /home/travis/build/ararslan/fftw-builder/test.jl, in expression starting on line 29
```
it exists, Julia just doesn't like it.

**Comment by tkelman** [comment]
it builds inside a docker container then tests outside right? the compiler inside the docker container probably wasn't built as multilib? I forget exactly how ubuntu spells it, but the package is something to the effect of libc6:i386 or glibc6:i386

**Comment by ararslan** [comment]
> it builds inside a docker container then tests outside right? the compiler inside the docker container probably wasn't built as multilib?

Ah yes, good call.

File: .travis.yml [hunk 4]
```diff
@@ -6,11 +6,16 @@ services:
 matrix:
   include:
     - os: linux
-      env: ARCH="i686" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x86"
+      env: ARCH="i686" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x86" JULIA_URL="https://julialangnightlies-s3.julialang.org/bin/linux/x86/julia-latest-linux32.tar.gz"
+      addons:
+        apt:
+          packages:
+            - libc6:i386
+            - zlib1g:i386
```
Comment (tkelman) [inline .travis.yml:14]: I'd think zlib should even depend on libc

Comment (ararslan) [reply .travis.yml:14]: Probably, but it doesn't hurt to be explicit

**Comment by ararslan** [comment]
Hm, still doesn't like libfftw3_threads.so on 32-bit Linux.

**Comment by ararslan** [comment]
Okay well the point of this PR was to get Windows working, and it does now, so I'll go ahead and merge this then figure out the Linux failure separately. I might try building 32-bit Linux on Circle since their VMs are actually i686.

**Pull request merged**